### DISCUSSION TITLE
Query Help

### DISCUSSION THREAD

**Discussion by clcarver** [mongodb/laravel-mongodb#2294] [Q&A]
Hello, I have been searching on how to accomplish this for the last few days so this is my last ditch effort to see if this is even possible. I am trying to build the following query using the query builder:

`
db.work_orders.aggregate([
    {
        '$match': {
                    '$and': [
            {"main_work_center": {'$in': ['WC_1', 'WC_2', 'WC_3', 'WC_4']}},
            {'statuses.system': {'$in':["REL", "CRTD"]}},
            {'statuses.system': {'$nin':["TECO", "CNF", "CLSD"]}},
            {'scheduled_start_date': {'$lte': 20210712}},
            {'scheduled_finish_date': {'$gte': 20210712}},
            {'$or' : [
                {'last_call.cycle_days': {'$gte': 30}},
                {'maintenance_plan': null}
            ]}
        ],
        },
    },
    {
    '$addFields': {
            'plannedWork': {'$sum': "$operations.planned_duration"}
        }
    },
    {
        '$match': {
            '$and': [
                { '$or' : [
                    {'plannedWork': {'$gte': 4}},
                    {'order_type': {'$in':["T1", "T3", "T6"]}}
                ]}
            ]
        }
    },
    {'$sort': {'main_work_center': 1}}
])`

This works great and I get the results from my database application.

I run into the issue translating to laravel:

`
        WorkOrder::whereIn('statuses.system', ['REL', 'CRTD'])
            ->whereNotIn('statuses.system', ['CNF'])
            ->whereNotIn('statuses.system', ['TECO', 'CNF', 'CLSD'])
            ->whereNotIn('statuses.user', ['CNCL'])
            ->where('main_work_center', ['WC_1', 'WC_2', 'WC_3', 'WC_4'])
            ->where('scheduled_start_date', '<=', 20210712)
            ->where('scheduled_finish_date', '>=', 20210712)
            ->where(function ($q) {
                $q->where('last_call.cycle_days', '>=', 30)
                    ->orWhereNull('maintenance_plan');
            })
            ->whereRaw(function ($collection) {
                return $collection->aggregate([
                    '$addFields' => [
                        'plannedWork' => ['$sum' => '$operations.planned_duration']
                    ]
                ]);
            })
            ->orderBy('main_work_center')
            ->paginate(10);
`

The whereRaw aggregate does not add the field to the returned results. Also adding the additional filters after adding the new field does not work either.

I would also be ok with just running that raw query directly if that is possible? Using the WorkOrder::raw() does not work.

Any help or pointing in the right direction would be great. Thank you!

**Comment by zigzagdev** [comment]
Hi! @clcarver ! 

I ran into a similar issue when trying to combine Laravel's query builder with MongoDB's aggregation pipeline. 
Sharing what worked for me below.

The core issue: `whereRaw()` is designed for filter conditions, not aggregation pipeline stages like `$addFields`. 
And `Model::raw()` without a closure just returns the collection instance without executing anything — which is why your approach didn't work maybe.

Therefore, to run an aggregation pipeline, pass a closure to `Model::raw()` and call `aggregate()` method on the collection directly.


### Example

```php
use App\Models\WorkOrder;

$pipeline = [
    [
        '$match' => [
            '$and' => [
                ['main_work_center' => ['$in' => ['WC_1', 'WC_2', 'WC_3', 'WC_4']]],
                ['statuses.system' => ['$in' => ['REL', 'CRTD']]],
                ['statuses.system' => ['$nin' => ['TECO', 'CNF', 'CLSD']]],
                ['scheduled_start_date' => ['$lte' => 20210712]],
                ['scheduled_finish_date' => ['$gte' => 20210712]],
                [
                    '$or' => [
                        ['last_call.cycle_days' => ['$gte' => 30]],
                        ['maintenance_plan' => null],
                    ],
                ],
            ],
        ],
    ],
    [
        '$addFields' => [
            'plannedWork' => ['$sum' => '$operations.planned_duration'],
        ],
    ],
    [
        '$match' => [
            '$or' => [
                ['plannedWork' => ['$gte' => 4]],
                ['order_type' => ['$in' => ['T1', 'T3', 'T6']]],
            ],
        ],
    ],
    [
        '$sort' => ['main_work_center' => 1],
    ],
];

$results = WorkOrder::raw(function ($collection) use ($pipeline) {
    return $collection->aggregate($pipeline);
});
```

This mirrors your original MongoDB shell query 1:1.

## About pagination

Aggregation pipelines don't work with `paginate()` out of the box, therfore you'll need to handle it manually following.

```php
$page = request()->get('page', 1);
$perPage = 10;
$skip = ($page - 1) * $perPage;

// Total count (without sort stage)
$countPipeline = array_merge(
    array_slice($pipeline, 0, -1),
    [['$count' => 'total']]
);

$totalResult = WorkOrder::raw(fn ($collection) => $collection->aggregate($countPipeline))->toArray();

$total = $totalResult[0]['total'] ?? 0;

// Paginated results
$paginatedPipeline = array_merge($pipeline, [
    ['$skip' => $skip],
    ['$limit' => $perPage],
]);

$results = WorkOrder::raw(fn ($collection) => $collection->aggregate($paginatedPipeline));

$pagination = new \Illuminate\Pagination\LengthAwarePaginator(
    iterator_to_array($results),
    $total,
    $perPage,
    $page,
    ['path' => request()->url(), 'query' => request()->query()]
);
```

### Converting to Eloquent models (optional)

If you need Eloquent model instances (for accessors/relationships):

```php
$models = collect($results)->map(fn ($document) => WorkOrder::newFromBuilder((array) $document));
```

Hope this helps! 🙌

**Comment by MaddipatlaChetan24** [comment]
Hi @clcarver ! Building on zigzagdev's answer above — there's a cleaner way to do this using the official `mongodb/laravel-mongodb` package's Aggregation Builder (available since v4.3+), and a way to fix the double-execution problem in the pagination approach.

Two improvements over the raw()/closure approach:

1. `Model::aggregate()->addRawStage(...)` lets you add your existing pipeline stages verbatim, using the package's actual supported API instead of the raw()+closure workaround (which the GitHub maintainers have an open issue about, since raw() returns a plain collection with no pagination support built in).
2. Using MongoDB's $facet stage, you can get your paginated data and your total count in one aggregate call instead of two — $facet runs both branches against the same already-filtered document set in a single pass, so you're not re-running your $match/$addFields stages twice.

Check your version first — you need mongodb/laravel-mongodb v4.3 or later:

```bash
composer show mongodb/laravel-mongodb
```

Full solution:

```php
use App\Models\WorkOrder;

$page = (int) request('page', 1);
$perPage = 10;
$skip = ($page - 1) * $perPage;

$pipeline = WorkOrder::aggregate()
    ->addRawStage('$match', [
        '$and' => [
            ['main_work_center' => ['$in' => ['WC_1', 'WC_2', 'WC_3', 'WC_4']]],
            ['statuses.system' => ['$in' => ['REL', 'CRTD']]],
            ['statuses.system' => ['$nin' => ['TECO', 'CNF', 'CLSD']]],
            ['scheduled_start_date' => ['$lte' => 20210712]],
            ['scheduled_finish_date' => ['$gte' => 20210712]],
            ['$or' => [
                ['last_call.cycle_days' => ['$gte' => 30]],
                ['maintenance_plan' => null],
            ]],
        ],
    ])
    ->addRawStage('$addFields', [
        'plannedWork' => ['$sum' => '$operations.planned_duration'],
    ])
    ->addRawStage('$match', [
        '$or' => [
            ['plannedWork' => ['$gte' => 4]],
            ['order_type' => ['$in' => ['T1', 'T3', 'T6']]],
        ],
    ])
    ->addRawStage('$sort', ['main_work_center' => 1])
    ->addRawStage('$facet', [
        'data' => [
            ['$skip' => $skip],
            ['$limit' => $perPage],
        ],
        'totalCount' => [
            ['$count' => 'count'],
        ],
    ]);

$result = $pipeline->get()->first();

$data = $result['data'] ?? [];
$total = $result['totalCount'][0]['count'] ?? 0;

$paginator = new \Illuminate\Pagination\LengthAwarePaginator(
    $data,
    $total,
    $perPage,
    $page,
    [
        'path' => request()->url(),
        'query' => request()->query(),
    ]
);

return $paginator;
```

This mirrors your original shell pipeline 1:1 in the $match/$addFields/$match/$sort stages, then adds one final $facet stage that produces both the current page's rows and the total count from a single execution.

If you need Eloquent model instances (for accessors or relationships) rather than plain arrays/BSON documents:

```php
$models = collect($data)->map(fn ($doc) => WorkOrder::newFromBuilder((array) $doc));

$paginator = new \Illuminate\Pagination\LengthAwarePaginator(
    $models,
    $total,
    $perPage,
    $page,
    ['path' => request()->url(), 'query' => request()->query()]
);
```

One caveat, to be upfront about it: I haven't run this against a live database myself — I verified addRawStage() exists via the package's own GitHub PR/changelog, but you should test it against your specific installed version before shipping it. If for any reason addRawStage() isn't available to you (older package version), the $facet single-call pagination trick still works fine inside the plain raw()/closure approach from the earlier answer — just replace the two separate WorkOrder::raw(...) calls (one for count, one for data) with a single pipeline that ends in the $facet stage shown above.
